hamilton/customization/api.py

```diff
@@ -0,0 +1,306 @@
+            task_id=task_id,
+        )
+
+    @abc.abstractmethod
+    def run_after_node_execution(
+        self,
+        *,
+        node_name: str,
+        node_tags: Dict[str, Any],
+        node_kwargs: Dict[str, Any],
+        node_return_type: type,
+        result: Any,
+        error: Optional[Exception],
+        success: bool,
+        task_id: Optional[str],
+        **future_kwargs: Any,
+    ):
+        """Hook that is executed post node execution.
+
+        :param node_name: Name of the node in question
+        :param node_tags: Tags of the node
+        :param node_kwargs: Keyword arguments passed to the node
+        :param node_return_type: Return type of the node
+        :param result: Output of the node, None if an error occurred
+        :param error: Error that occurred, None if no error occurred
+        :param success: Whether the node executed successfully
+        :param task_id: The ID of the task, none if not in a task-based environment
+        :param future_kwargs: Additional keyword arguments -- this is kept for backwards compatibility
+        """
+
+    def post_node_execute(
+        self,
+        *,
+        run_id: str,
+        node_: node.Node,
+        kwargs: Dict[str, Any],
+        success: bool,
+        error: Optional[Exception],
+        result: Optional[Any],
+        task_id: Optional[str] = None,
+    ):
+        """Wraps the after_execution method, providing a bridge to an external-facing API. Do not override this!"""
+        self.run_after_node_execution(
+            node_name=node_.name,
+            node_tags=node_.tags,
+            node_kwargs=kwargs,
+            node_return_type=node_.type,
+            result=result,
+            error=error,
+            task_id=task_id,
+            success=success,
+        )
+
+
+class EdgeConnectionHook(BaseDoCheckEdgeTypesMatch, BaseDoValidateInput, abc.ABC):
+    def do_check_edge_types_match(self, *, type_from: type, type_to: type) -> bool:
+        """Wraps the check_edge_types_match method, providing a bridge to an external-facing API. Do not override this!"""
+        return self.check_edge_types_match(type_from, type_to)
+
+    @abc.abstractmethod
+    def check_edge_types_match(self, type_from: type, type_to: type, **kwargs: Any) -> bool:
+        """This is run to check if edge types match. Note that this is an OR functionality
+        -- this is run after we do some default checks, so this can only be permissive.
+        Reach out if you want to be more restrictive than the default checks.
+
+        :param type_from: The type of the node that is the source of the edge.
+        :param type_to: The type of the node that is the destination of the edge.
+        :param kwargs: This is kept for future backwards compatibility.
+        :return: Whether or not the two node types form a valid edge.
+        """
+        return self.do_check_edge_types_match(type_from=type_from, type_to=type_to)
```

Comment:
don't these two functions refer to each other? I'm confused.

Reply:
Yeah this was a bug -- hadn't been using this class